### PR TITLE
Add `extension_options` parameter to `cythonize()`

### LINKED ISSUE #1480: docs error: cythonize(include_path=x) does not set C include path
The documentation in [compilation.rst](../blob/master/docs/src/reference/compilation.rst#configuring-the-c-build) claims that

```
cythonize(extensions, include_path=[numpy.get_include()])
```

adds the numpy include path to the include file search path used by the compiler. This has been quoted many times and the documented feature is used by many packages, although with zero effect: `include_path` only adds to the search path used by Cython to find `pyx` and `pxd` files.

Some alternative ways to address this:
1. Change docs to match code by removing the misleading section.
2. Change code to match docs by adding the feature. 
3. Change docs and code to use `cython(include_dirs=[])` as a way for globally specifying C include paths. 
4. Change it all to have `cython_include_dirs` and `c_include_dirs` (or path).

3 has the benefit that code using `include_path` will continue to function as before (no effect on C include path), while making the feature people have obviously been expecting available. The term `include_dir` would then consistently (albeit not obviously) refer to the C includes and the term `include_path` to the Cython includes.

### PR DESCRIPTION
**Pull request by matusvalo**
This PR is a follow up of #1512. It rebases main and do a small refactoring + adds unittests.

Fixes #1480

### CODE REVIEW

File: docs/src/userguide/source_files_and_compilation.rst [hunk 1]
```diff
@@ -157,10 +157,18 @@ If you have include files in non-standard places you can pass an
...
+objects. Check the `distutils documentation`_ for a description of
+possible options.  Some useful options to know about
+are ``include_dirs``, ``libraries``, and ``library_dirs`` which specify where
+to find the ``.h`` and library files when linking to external libraries.
+
 Often, Python packages that offer a C-level API provide a way to find
 the necessary include files, e.g. for NumPy::
 
-    include_path = [numpy.get_include()]
+    extension_options = {'include_path': numpy.get_include()}
```
Comment (dalcinl) [inline docs/src/userguide/source_files_and_compilation.rst:171]: AFAICT, `include_path` is not a distutils option. I believe you meant `include_dirs` in this example.

Comment (matusvalo) [reply docs/src/userguide/source_files_and_compilation.rst:171]: Yes, you are right: https://setuptools.pypa.io/en/latest/userguide/ext_modules.html Thanks!

**Comment by dalcinl** [comment]
IMHO, all the 8 commits in this PR should be all squashed in a single one. I don't see the value of keeping such details in the git history. For example, the commits started adding a `ext_opts` option that a few commits after gets renamed.

**Comment by da-woods** [comment]
> IMHO, all the 8 commits in this PR should be all squashed in a single one. I don't see the value of keeping such details in the git history. For example, the commits started adding a `ext_opts` option that a few commits after gets renamed.

We usually just squash it when we merge it so it so it isn't a big concern.

**Comment by 0dminnimda** [comment]
Yep, usually it's better to just squish and merge the pr.
With a steady stream of PRs it's easier to see what commits relate to what features + it's probably easier to trace the bugs

**Comment by dalcinl** [comment]
> Why is this better than creating an `Extension` object, maybe with a wildcard extension name, and passing that into `cythonize()`?

Are you sure the full contents of a Extension instance with wildcard sources will propagate to expanded Extension objects? The code in here is a bit long/convoluted, but I'm looking at the `default_create_extension()` function, and looks like the Extension template's `__dict__` contents are not copied over to the newly create Extension object.

PS: On the other hand, with core Python sunsetting distutils in the near future, maybe Cython should stop extending all this distutils-based stuff?

**Comment by scoder** [comment]
>with core Python sunsetting distutils in the
>near future, maybe Cython should stop extending all this distutils-based stuff?

Yeah, Cython 3.0 seems the right time now to switch completely to setuptools.

**Comment by scoder** [comment]
>Are you sure the full contents of a Extension instance with wildcard
>sources will propagate to expanded Extension objects?

If that turns out to be not the case, then that might be a bug worth exploring. But the fix for it is certainly not to create a new, completely separate API option. Especially now that there is a unique Extension class again that we can simply refer our users to.

**Comment by scoder** [comment]
Thank you for your efforts, Matus, and sorry for the initial confusion that made you workk on this. I don't think this is something we should pursue. If there is anything wrong with passing `Extension` objects into `cythonize()`, then we should fix that instead.

**Pull request closed without merging**